File: code/ts/modular_api/src/core/tracing/w3cTraceContextPropagator.ts
import {
  createTraceState,
  isValidSpanId,
  isValidTraceId,
  TraceFlags,
  trace,
  type Context,
  type TextMapGetter,
  type TextMapPropagator,
  type TextMapSetter,
} from '@opentelemetry/api';

/** The W3C `traceparent` carrier key. */
export const TRACEPARENT_HEADER = 'traceparent';

/** The W3C `tracestate` carrier key. */
export const TRACESTATE_HEADER = 'tracestate';

/** The only version defined by the W3C Trace Context recommendation. */
const SUPPORTED_VERSION = '00';

/** `00-<32 hex>-<16 hex>-<2 hex>` — lowercase only, every field fixed-width. */
const TRACEPARENT_PATTERN = /^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/;

/**
 * Propagates W3C Trace Context: `traceparent` and `tracestate`.
 *
 * This is ours because `@opentelemetry/api` does not ship a propagator — the W3C
 * implementation lives in `@opentelemetry/core`, which is past the API boundary
 * that ADR-0005 A2 draws and which the Stage 1b dependency guard forbids
 * (runbook D26). Python's official API package includes an equivalent, and that
 * implementation is the reference these cases are checked against.
 *
 * The format is exact: 55 characters, `00-<trace id>-<parent span id>-<flags>`,
 * lowercase hex only, version `00` the only one defined. Anything else is
 * rejected, and rejection means *return the context unchanged* — a broken header
 * from an upstream caller must never fail a request.
 *
 * `tracestate` is carried verbatim and never interpreted (runbook D4). This
 * implementation is more faithful to that than the Dart one can be: the
 * JavaScript API's `createTraceState` accepts the raw header string, so ordering
 * and formatting survive untouched, whereas the Dart API only accepts a map.
 *
 * The Dart counterpart is `lib/src/core/tracing/w3c_trace_context_propagator.dart`.
 */
export class W3CTraceContextPropagator implements TextMapPropagator<unknown> {
  fields(): string[] {
    return [TRACEPARENT_HEADER, TRACESTATE_HEADER];
  }

  inject(context: Context, carrier: unknown, setter: TextMapSetter<unknown>): void {
    const spanContext = trace.getSpanContext(context);
    if (spanContext === undefined) return;
    if (!isValidTraceId(spanContext.traceId) || !isValidSpanId(spanContext.spanId)) return;

    const flags = (spanContext.traceFlags & TraceFlags.SAMPLED) === TraceFlags.SAMPLED ? '01' : '00';
    setter.set(
      carrier,
      TRACEPARENT_HEADER,
      `${SUPPORTED_VERSION}-${spanContext.traceId}-${spanContext.spanId}-${flags}`,
    );

    // Only when there is something to carry. An empty tracestate is a header that
    // says nothing.
    const serialized = spanContext.traceState?.serialize();
    if (serialized !== undefined && serialized.length > 0) {
      setter.set(carrier, TRACESTATE_HEADER, serialized);
    }
  }

  extract(context: Context, carrier: unknown, getter: TextMapGetter<unknown>): Context {
    try {
      const raw = first(getter.get(carrier, TRACEPARENT_HEADER));
      if (raw === undefined || raw.length === 0) return context;

      const match = TRACEPARENT_PATTERN.exec(raw);
      if (match === null) return context;

      const [, version, traceId, spanId, flagsHex] = match;

      // Version 00 only. A future version may extend the layout, and guessing at
      // an unknown one is worse than starting a fresh trace.
      if (version !== SUPPORTED_VERSION) return context;
      if (!isValidTraceId(traceId) || !isValidSpanId(spanId)) return context;

      // Bit 0 is `sampled`. Unknown bits must not defeat the one we understand.
      const flags = Number.parseInt(flagsHex, 16);
      const sampled = (flags & TraceFlags.SAMPLED) === TraceFlags.SAMPLED;

      const rawState = first(getter.get(carrier, TRACESTATE_HEADER));

      return trace.setSpanContext(context, {
        traceId,
        spanId,
        traceFlags: sampled ? TraceFlags.SAMPLED : TraceFlags.NONE,
        isRemote: true,
        // Verbatim: the raw header goes in untouched (D4).
        traceState:
          rawState !== undefined && rawState.trim().length > 0
            ? createTraceState(rawState)
            : undefined,
      });
    } catch {
      return context;
    }
  }
}

/** This API's getter may return a single value or a list. */
function first(value: undefined | string | string[]): string | undefined {
  return Array.isArray(value) ? value[0] : value;
}

File: code/dart/modular_api/lib/src/core/tracing/w3c_trace_context_propagator.dart
import 'package:dartastic_opentelemetry_api/dartastic_opentelemetry_api.dart';

/// The W3C `traceparent` carrier key.
const String traceparentHeader = 'traceparent';

/// The W3C `tracestate` carrier key.
const String tracestateHeader = 'tracestate';

/// The only version defined by the W3C Trace Context recommendation.
const String _supportedVersion = '00';

/// `00-<32 hex>-<16 hex>-<2 hex>` — lowercase only, every field fixed-width.
final RegExp _traceparentPattern =
    RegExp(r'^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$');

final String _invalidTraceId = '0' * 32;
final String _invalidSpanId = '0' * 16;

/// Propagates W3C Trace Context: `traceparent` and `tracestate`.
///
/// This is ours because the Dart OpenTelemetry **API** does not ship a
/// propagator — the W3C implementation lives in the SDK, which core must not
/// depend on (ADR-0005 A2, runbook D26). Python's official API package includes
/// an equivalent, which is both the argument that this belongs at the API layer
/// and the reference our behaviour is checked against.
///
/// The format is exact: 55 characters, `00-<trace id>-<parent span id>-<flags>`,
/// lowercase hex only, version `00` the only one defined. Anything else is
/// rejected, and rejection means *return the context unchanged* — a broken header
/// from an upstream caller must never fail a request.
///
/// `tracestate` is carried verbatim and never interpreted (runbook D4). We are not
/// a tracing vendor, so we add no entry of our own; normalising or reordering what
/// another vendor set would corrupt information we do not own.
///
/// Header-name casing is not this class's concern. HTTP header semantics belong to
/// the carrier, so [fields] reports lowercase canonical names and lookup goes
/// through the supplied [TextMapGetter]. Shelf already lowercases header names.
class W3CTraceContextPropagator<C> implements TextMapPropagator<C, String> {
  /// Creates a W3C Trace Context propagator.
  const W3CTraceContextPropagator();

  @override
  List<String> fields() =>
      const <String>[traceparentHeader, tracestateHeader];

  @override
  void inject(
    Context context,
    C carrier,
    TextMapSetter<String> setter,
  ) {
    final spanContext = context.spanContext;
    if (spanContext == null || !spanContext.isValid) return;

    final flags = spanContext.traceFlags.isSampled ? '01' : '00';
    setter.set(
      traceparentHeader,
      '$_supportedVersion-${spanContext.traceId.hexString}'
      '-${spanContext.spanId.hexString}-$flags',
    );

    // Only when there is something to carry. Emitting an empty tracestate would
    // be a header that says nothing.
    final traceState = spanContext.traceState;
    if (traceState != null) {
      final encoded = traceState.toString();
      if (encoded.isNotEmpty) setter.set(tracestateHeader, encoded);
    }
  }

  @override
  Context extract(
    Context context,
    C carrier,
    TextMapGetter<String> getter,
  ) {
    try {
      final raw = getter.get(traceparentHeader);
      if (raw == null || raw.isEmpty) return context;

      final match = _traceparentPattern.firstMatch(raw);
      if (match == null) return context;

      // Version 00 only. A future version may extend the format, and guessing at
      // an unknown layout is worse than starting a fresh trace.
      if (match.group(1) != _supportedVersion) return context;

      final traceId = match.group(2)!;
      final spanId = match.group(3)!;
      if (traceId == _invalidTraceId || spanId == _invalidSpanId) return context;

      // Bit 0 is `sampled`. Unknown bits must not defeat the one we understand.
      final flags = int.parse(match.group(4)!, radix: 16);
      final sampled = (flags & TraceFlags.SAMPLED_FLAG) == TraceFlags.SAMPLED_FLAG;

      return context.withSpanContext(
        OTelAPI.spanContext(
          traceId: OTelAPI.traceIdFrom(traceId),
          spanId: OTelAPI.spanIdFrom(spanId),
          traceFlags: OTelAPI.traceFlags(
            sampled ? TraceFlags.SAMPLED_FLAG : TraceFlags.NONE_FLAG,
          ),
          traceState: _extractTraceState(getter),
          isRemote: true,
        ),
      );
    } catch (_) {
      return context;
    }
  }

  /// Reads `tracestate` without parsing its grammar (D4).
  ///
  /// The value is split into entries only because [TraceState] is the shape the
  /// API stores; nothing is validated, reordered or dropped.
  TraceState? _extractTraceState(TextMapGetter<String> getter) {
    final raw = getter.get(tracestateHeader);
    if (raw == null || raw.trim().isEmpty) return null;

    final entries = <String, String>{};
    for (final member in raw.split(',')) {
      final separator = member.indexOf('=');
      if (separator <= 0) continue;
      entries[member.substring(0, separator).trim()] =
          member.substring(separator + 1).trim();
    }

    return entries.isEmpty ? null : OTelAPI.traceState(entries);
  }
}
